draft-ietf-quic-transport.md

```diff
@@ -732,26 +771,44 @@ sequence of frames, as described in {{frames}}.
-using the Omit Connection ID flag.  When present, the Connection ID is in the
-same location in all packet headers, making it straightforward for middleboxes,
-such as load balancers, to locate and use it.
-
-The client MUST choose a random connection ID and use it in Initial packets
-({{packet-initial}}) and 0-RTT packets ({{packet-protected}}).
-
-When the server receives a Initial packet and decides to proceed with the
-handshake, it chooses a new value for the connection ID and sends that in a
-Retry ({{packet-retry}}) or Handshake ({{packet-handshake}}) packet.  The server
-MAY choose to use the value that the client initially selects.
-
-Once the client receives the connection ID that the server has chosen, it MUST
-use it for all subsequent Handshake ({{packet-handshake}}) and 1-RTT
-({{packet-protected}}) packets but not for 0-RTT packets ({{packet-protected}}).
-
-Server's Version Negotiation ({{packet-version}}) and Retry ({{packet-retry}})
-packets MUST use connection ID selected by the client.
+A connection ID is used to ensure consistent routing of packets.  The long
+header contains two connection IDs: the Destination Connection ID is chosen by
+the recipient of the packet and is used to provide consistent routing; the
+Source Connection ID is used to set the Destination Connection ID used by the
+peer.
+
+During the handshake, packets with the long header are used to establish the
+connection ID that each endpoint uses.  Each endpoint uses the Source Connection
+ID field to specify the connection ID that is used in the Destination Connection
+ID field of packets being sent to them.  Upon receiving a packet, each endpoint
+sets the Destination Connection ID it sends to match the value of the Source
+Connection ID that they receive.
+
+During the handshake, an endpoint might receive multiple packets with the long
+header, and thus be given multiple opportunities to update the Destination
+Connection ID it sends.  A client MUST only change the value it sends in the
+Destination Connection ID in response to the first packet of each type it
+receives from the server (Retry or Handshake); a server MUST set its value based
+on the Initial packet.  Any additional changes are not permitted; subsequent
+packets of those types include a different Source Connection ID, they MUST be
+discarded.  This avoids problems that might arise from stateless processing of
+multiple Initial packets producing different connection IDs.
+
+Short headers only include the Destination Connection ID and omit the explicit
+length.  The length of the Destination Connecton ID field is expected to be
+known to endpoints.  Endpoints with a load balancer that uses a connection ID
+can agree to a fixed or minimum length for connection IDs with necessary
+information in the fixed portion.  If that fixed portion instead encodes an
+explicit length, the entire connection ID can vary in length and still be used
+by the load balancer.
```

Comment:
This part about load balancers is confusing. I would at least make it a separate paragraph. I do wonder if this is necessary ... it's basically a suggestion for how load balancers are implemented, and I would be more explicit about that in this para. Here's a suggested rephrasing:

"Endpoints using a connection-ID based load balancer could agree with the load balancer on a fixed or minimum length and on an encoding for connection IDs. This fixed portion could encode an explicit length, which allows the entire connection ID to vary in length and still be used by the load balancer."